src/coreclr/vm/CachedInterfaceDispatchPal.h

```diff
@@ -0,0 +1,185 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#ifndef __CACHEDINTERFACEDISPATCHPAL_H__
+#define __CACHEDINTERFACEDISPATCHPAL_H__
+
+#ifdef FEATURE_CACHED_INTERFACE_DISPATCH
+
+extern "C" void RhpInitialInterfaceDispatch();
+
+
+bool InterfaceDispatch_InitializePal();
+
+// Allocate memory aligned at sizeof(void*)*2 boundaries
+void *InterfaceDispatch_AllocDoublePointerAligned(size_t size);
+// Allocate memory aligned at sizeof(void*) boundaries
+
+void *InterfaceDispatch_AllocPointerAligned(size_t size);
+
+enum Flags
+{
+    // The low 2 bits of the m_pCache pointer are treated specially so that we can avoid the need for
+    // extra fields on this type.
+    // OR if the m_pCache value is less than 0x1000 then this is a vtable offset and should be used as such
+    IDC_CachePointerPointsIsVTableOffset = 0x2,
+    IDC_CachePointerPointsAtCache = 0x0,
+    IDC_CachePointerMask = 0x3,
+    IDC_CachePointerMaskShift = 0x2,
+};
+
+enum class DispatchCellType
+{
+    InterfaceAndSlot = 0x0,
+    VTableOffset = 0x2,
+};
+
+struct DispatchCellInfo
+{
+private:
+    static DispatchCellType CellTypeFromToken(DispatchToken token)
+    {
+        if (token.IsThisToken())
+        {
+            return DispatchCellType::VTableOffset;
+        }
+        return DispatchCellType::InterfaceAndSlot;
+    }
+public: 
+
+    DispatchCellInfo(DispatchToken token, bool hasCache) :
+        CellType(CellTypeFromToken(token)),
+        Token(token),
+        HasCache(hasCache ? 1 : 0)
+    {
+
+    }
+    const DispatchCellType CellType;
+    const DispatchToken Token;
+
+    uintptr_t GetVTableOffset() const
+    {
+        if (CellType == DispatchCellType::VTableOffset)
+        {
+            uint32_t slot = Token.GetSlotNumber();
+            unsigned offsetOfIndirection = MethodTable::GetVtableOffset() + MethodTable::GetIndexOfVtableIndirection(slot) * TARGET_POINTER_SIZE;
+            unsigned offsetAfterIndirection = MethodTable::GetIndexAfterVtableIndirection(slot) * TARGET_POINTER_SIZE;
+
+            uintptr_t offsetOfIndirectionPortion = (((uintptr_t)offsetOfIndirection) << ((TARGET_POINTER_SIZE * 8) / 2));
+            uintptr_t offsetAfterIndirectionPortion = (((uintptr_t)offsetAfterIndirection) << ((TARGET_POINTER_SIZE * 8) / 4));
+            uintptr_t flagPortion = (uintptr_t)IDC_CachePointerPointsIsVTableOffset;
+
+            uintptr_t result = offsetOfIndirectionPortion | offsetAfterIndirectionPortion | flagPortion;
+            _ASSERTE(slot == VTableOffsetToSlot(result));
+            return result;
+        }
+        return 0;
+    }
+
+    static unsigned VTableOffsetToSlot(uintptr_t vtableOffset)
+    {
+        unsigned offsetOfIndirection = (unsigned)(vtableOffset >> ((TARGET_POINTER_SIZE * 8) / 2));
+        unsigned offsetAfterIndirection = (unsigned)(vtableOffset >> ((TARGET_POINTER_SIZE * 8) / 4)) & 0xFF;
```

Comment:
Why is this being masked to 8 bits? Might be good to have a comment, it's not evident from staring at this file